flang/lib/Semantics/semantics.cpp

```diff
@@ -114,6 +114,11 @@ class SemanticsVisitor : public virtual BaseChecker, public virtual C... {
     context_.set_location(std::nullopt);
   }
 
+  // This is necessary to avoid "walking" into the Fail clause,
+  // which confuses the CheckAllowed into thinking there's another
+  // memoryorder, when it's actually the argument to the fail clause.
+  bool Pre(const parser::OmpFailClause &) { return false; }
```

Comment:
Would this disallow any future checks for the OmpFailClause?

What are the alternatives here?
Marking the MemoryOrder values for OmpFail as an ENUM or something?
Alternative, would be to manually check the clauses having MemoryOrder components in check-omp-structure.cpp.

Reply:
Assuming we have the correct parser, there is no way that the FAIL clause will contain anything other than a MemoryOrder. 

The problem we're trying to solve here is that the default is to `Walk` everything where `Pre` returns true.  So if we have `!$OMP ATOMIC COMPARE SEQ_CST FAIL(RELAXED)`, it will get the the FAIL, and default `Pre` returns true, so it gets to process the RELAXED. It treats that as if it was part of the regular list of memory order objects listed, and says "There's two memory-orders on the same ATOMIC COMPARE" (not the precise wording). 

We COULD use a different type, and then have a separate checker bit, but than we will also end up with two (almost) identical parsers for memory order objects, and I'm not sure exactly what it solves to do that. 

We can always call some function inside `Pre` that checks things. Or call check the fail clause in the check-omp-structure.cpp. 

Really, none of the solutions are fantastic. This works for the moment, and I'm not sure it makes sense to add more complexity, unless we have a good reason for that.

Reply:
I was asking whether this solution will prevent the ability to perform the following semantic check from the OpenMP 5.2 standard.

-> `acq_rel` and `release` cannot be specified as arguments to the `fail` clause.

If you do not want to change the parse-tree representation, an alternative solution could be to remember whether we are visiting the FailClause and only call `CheckAllowedClause` for the MemoryOrder clauses if we are not inside  the FailClause. Something like the following (only done for Release MemoryOrder clause).

```
+void OmpStructureChecker::Enter(const parser::OmpFailClause &) {
+  CheckAllowedClause(llvm::omp::Clause::OMPC_fail);
+  isFailClause = true;
+}
+
+void OmpStructureChecker::Leave(const parser::OmpFailClause &) {
+  isFailClause = false;
+}
+
+void OmpStructureChecker::Enter(const parser::OmpClause::Release &) {
+  if(!isFailClause) {
+    CheckAllowedClause(llvm::omp::Clause::OMPC_release);
+  }
+}
```

Note: The above code will also involve removing the CHECK_SIMPLE_CLAUSE entries for the Release MemoryOrder clause, adding isFailClause as a member of OmpStructureChecker, and adding the Enter and Leave signatures in OmpStructureChecker for the OmpFailClause.

Reply:
Ah, I see. As far as I can tell, there's nothing automatic (in OMP.td) that provides a help for the fail clause's memory order options. We'd have to write code for that somewhere. We already do a bunch of checking for memoryorder in the semantics of ATOMIC in general, and we also need to check that the fail clause doesn't appear outside of a COMPARE (at least I think that's how it's meant to work).

The alternative suggested would be a workable solution. In fact, I'm liking that more, mostly because it doesn't alter a generic class used everywhere. I'll implement that.

In both cases, we still need to know whether we're checking a memory order inside a fail clause or not - and I guess we would add an `else Say("Release memory order not allowed on FAIL clause")` in the not allowed cases (not exactly how it would be written, but conceptually).